File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
       
        ListNode* curr = head;
        int c=0;
        
        while(curr != NULL){
            curr=curr->next;
            c++;
        }
        
        int gap = c-n;
        
        if(gap == 0){
            ListNode* node = head;
            ListNode* newHead = head->next;
            delete(node);
            return newHead;
        }
        
        curr = head;
        
        for(int i=1; i<gap; i++){
            curr = curr->next;
        }
        
        ListNode* node = curr->next;
        curr->next = curr->next->next;
        delete(node);
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
        ListNode dh = new ListNode(-1);
        dh.next = head;
        
        ListNode left = dh;
        ListNode right = head;
        
        while(n>0 && right != null){
            right = right.next;
            n--;
        }
        
        while(right != null){
            left = left.next;
            right = right.next;
        }
        
        left.next = left.next.next;
        return dh.next;
        
        
    }
}